### PR TITLE
Pyspark sampling function

### PR DESCRIPTION
**Pull request by AndreSchumacher**
This is more of a draft than an actual Pull Request. In the JIRA Josh was proposing to fall back to Python's native random in case NumPy is not installed, while preserving the results. I find that hard to implement given that the random generators used may be very different, although one could emulate drawing values from a Poisson distribution via random.expovariate() I guess. So I was wondering whether we could discuss this here.

Also note that takeSample() is still missing. I wanted to discuss sample() first before coming to that.
Update: added takeSample() based on RDD.scala

JIRA: https://spark-project.atlassian.net/browse/SPARK-651

### CODE REVIEW

**Comment by mateiz** [comment]
Hey Andre,

I agree with Josh that we should fall back on the default if numpy isn't available. I suggest to test whether numpy is available on the master first. If it is, always try to use it on the workers -- otherwise, always use the system default. If someone installs numpy on some workers but not others, or on the master but not workers, that's their fault.

**Comment by AndreSchumacher** [comment]
OK, falling back on the default random.py when numpy is unavailable sounds good (as long as people don't expect the results to be the same).

Is there any more practical way to pass settings from master to workers than serializing it into its input stream, using SparkEnv maybe?

Thanks for the comments!

**Comment by AndreSchumacher** [comment]
So I refactored things a bit. The whole drawing of random numbers now happens inside RDDSampler, which determines which of the corresponding functions to call. Based on a little experiment I did comparing the two Poisson generators they seem to come close.

If I understand correctly the RDDSampler should actually be created on the master when it's constructor is invoked. From them on the decision which of the libraries to use should be consistent among the workers. Would be great if someone who knows the code better could comment on that.

File: python/pyspark/rdd.py [hunk 1]
```diff
@@ -165,14 +167,61 @@ def distinct(self):
...
+        fraction = 0.0
+        total = 0
+        multiplier = 3.0
+        # TODO: it seems expensive to start a job just to initialize this variable but it may
+        # not be possible to avoid it (Spark does the same)
+        initialCount = self.count()
+        maxSelected = 0
+
+        if (num < 0):
+            raise Exception("Negative number of elements requested")
```
Comment (JoshRosen) [inline python/pyspark/rdd.py]: I'd raise [ValueError](http://docs.python.org/2/library/exceptions.html#exceptions.ValueError) here.

File: python/pyspark/rdd.py [hunk 2]
```diff
@@ -165,14 +167,61 @@ def distinct(self):
...
+        else:
+            fraction = multiplier * (num + 1) / initialCount
+            total = num
+
+        samples = self.sample(withReplacement, fraction, seed).collect()
+    
+        # If the first sample didn't turn out large enough, keep trying to take samples;
+        # this shouldn't happen often because we use a big multiplier for their initial size
+        while len(samples) < total:
+            if seed > sys.maxint -2:
```
Comment (JoshRosen) [inline python/pyspark/rdd.py]: There should be an extra space next to this minus sign.

**Comment by AndreSchumacher** [comment]
Disabled the doctest tests for sample() and takeSample() for now since they are bound to fail in numpy-installed versus non-numpy cases.

**Comment by AndreSchumacher** [comment]
Jenkins, retest this please

**Comment by mateiz** [comment]
Hey Andre, I tried this out and it looks good. Thanks!

**Pull request merged**